### DISCUSSION TITLE
Make the firmware run faster

### DISCUSSION THREAD

**Discussion by xkhainguyen** [bitcraze/discussions#757] [Q&A] [answered]
Hi, 

I am poking at the build system of CF2.1 firmware and I wonder if it is possible to make everything run faster for my purpose. There are some I can think of but I don't know if it can break anything.

1. General compile flag from -Os to -O3 (https://github.com/bitcraze/crazyflie-firmware/blob/78b2f42f9554b309fd987726327fbc1f59788d84/Makefile#LL114C1-L118C6)
2. Turn on CMSIS-DSP options (NEON, OPTIMIZED, AUTOVECTORIZE, etc)
3. Remove any ASSERT, error checking, etc (assume everything is working fine)

Could please give me some insights about this?
Thank you very much!

**Comment by krichardsson** [comment, marked as answer]
Hi, 

> General compile flag from -Os to -O3 (https://github.com/bitcraze/crazyflie-firmware/blob/78b2f42f9554b309fd987726327fbc1f59788d84/Makefile#LL114C1-L118C6)

You can try. I don't think it will make a huge difference but it might be worth a test. This should not (assuming the code is well written) cause any problems, as far as I know. 

> Turn on CMSIS-DSP options (NEON, OPTIMIZED, AUTOVECTORIZE, etc)

I don't think there is that much code that would benefit from this, but again you can try

> Remove any ASSERT, error checking, etc (assume everything is working fine)

Easy to try, I don't think it will make a big difference either.

I assume you need to free CPU power for other purposes? In general we try to be fairly conservative in CPU usage and I don't think there are any obvious parts that waste CPU power. You can start to remove functionality from the firmware, but most of the functionality that is not used does not cost in CPU either. 

The area I would target is probably the kalman estimator (if you are using it). You can maybe look at reducing the data that is sent to it, especially for the lighthouse.

**Reply by xkhainguyen** [reply]
Thank you very much!
I try to use -O3 and everything is a little bit faster.
